### ISSUE TITLE
mlaunch assumes replica set initialized when retries run out

### ISSUE THREAD

**Issue by p-mongo** [rueckstiess/mtools#858]
mlaunch has logic to retry replica set initialization:

```
    def _initiate_replset(self, port, name, maxwait=30):
        """Initiate replica set."""
        if not self.args['replicaset'] and name != 'configRepl':
            if self.args['verbose']:
                print('Skipping replica set initialization for %s' % name)
            return

        con = self.client('localhost:%i' % port)
        try:
            rs_status = con['admin'].command({'replSetGetStatus': 1})
            return rs_status
        except OperationFailure:
            # not initiated yet
            for i in range(maxwait):
                try:
                    con['admin'].command({'replSetInitiate':
                                          self.config_docs[name]})
                    break
                except OperationFailure as e:
                    print(e.message + " - will retry")
                    time.sleep(1)

            if self.args['verbose']:
                print("initializing replica set '%s' with configuration: %s"
                      % (name, self.config_docs[name]))
            print("replica set '%s' initialized." % name)

```

This logic retries up to the specified number of times, and when all of the retries have been performed, assumes success. However every attempt may have failed, in which case the replica set remains unititialized.

### Expected behavior

mlaunch should exit with an error if it is unable to initialize the replica set.

### Actual/current behavior

mlaunch claims success,  application fails at a later time because replica set does not actually exist.

```
+ python2 -m mtools.mlaunch.mlaunch --dir /tmpfs/db --binarypath /opt/mongodb/bin --setParameter enableTestCommands=1 --setParameter acceptApiVersion2=1 --setParameter diagnosticDataCollectionEnabled=false --replicaset --name test-rs --nodes 2 --arbiter
launching: "/opt/mongodb/bin/mongod" on port 27017
launching: "/opt/mongodb/bin/mongod" on port 27018
launching: "/opt/mongodb/bin/mongod" on port 27019
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
Unsupported OP_QUERY command: replSetInitiate - will retry
replica set 'test-rs' initialized.
# ...
+ ruby -Ilib -I.evergreen/lib -rbundler/setup -rserver_setup -e ServerSetup.new.setup_tags
/app/lib/mongo/server_selector/base.rb:285:in `select_server': No primary_preferred server is available in cluster: #<Cluster topology=Unknown[localhost:27017] servers=[#<Server address=localhost:27017 GHOST>]> with timeout=30.0, LT=0.015 (Mongo::Error::NoServerAvailable)
	from /app/lib/mongo/cluster.rb:975:in `validate_session_support!'
	from /app/lib/mongo/client.rb:1171:in `get_session!'
	from /app/lib/mongo/client.rb:1062:in `get_session'
	from /app/lib/mongo/client.rb:1088:in `with_session'
	from /app/lib/mongo/database.rb:225:in `command'
	from /opt/rubies/ruby-3.1.0/lib/ruby/3.1.0/forwardable.rb:238:in `command'
	from /app/.evergreen/lib/server_setup.rb:15:in `setup_tags'
	from -e:1:in `<main>'
/home/w/apps/ruby-driver/spec/shared/lib/mrss/docker_runner.rb:168:in `run_command': Process exited with code 1 (RuntimeError)
	from /home/w/apps/ruby-driver/spec/shared/lib/mrss/docker_runner.rb:117:in `run_tests'
	from /home/w/apps/ruby-driver/spec/shared/lib/mrss/docker_runner.rb:33:in `run'
	from ./.evergreen/test-on-docker:11:in `<main>'
```

### Steps to reproduce the actual/current behavior

Launch a 5.2 replica set.

### Environment

| Software         | Version
| ---------------- | -------
| mtools           | 1.7.0
| Python           | 3.7
| MongoDB server   | 5.2.1
| Operating system | Linux Debian testing

**Comment by p-mongo** [comment]
Looking at the bottom of the quoted code snippet, the configuration is reported after the replica set is initialized - shouldn't it be reported before? And why are both configuration and success only reported when the first attempt to initialize RS failed?

**Comment by stennie** [comment]
Definitely a bug. Will fix via #858 which you also reported.

Thanks,
Stennie

**Comment by stennie** [comment]
> Looking at the bottom of the quoted code snippet, the configuration is reported after the replica set is initialized - shouldn't it be reported before? 

I think the original message may possibly have been intended as "init is done" rather than "init is starting", but the text of the message suggests "init is starting". I applied your suggested patch from #868.

Cheers,
Stennie
